libs/common/src/sql-db-entities/payment/payment.entity.ts

```diff
@@ -51,4 +54,26 @@ export class PaymentEntity extends AuditableBaseEntity {
     payment.updateLastUpdatedBy(requestSource);
     return payment;
   }
+
+  public feeRecordsMatchPayment(): boolean {
+    if (this.feeRecords.length === 0) {
+      throw new PaymentHasNoFeeRecordsError();
+    }
+
+    if (this.feeRecords.some(({ paymentCurrency }) => paymentCurrency !== this.currency)) {
+      throw new PaymentAndFeeRecordCurrencyDoesNotMatchError();
+    }
+
+    const tolerance = PAYMENT_MATCHING_CURRENCY_TO_TOLERANCE_MAP[this.currency];
+
+    const totalFeesPaidToUkefForThePeriodInPaymentCurrency = this.feeRecords
+      .map((feeRecord) => feeRecord.getFeesPaidToUkefForThePeriodInThePaymentCurrency())
+      .reduce((total, feesPaid) => total.add(feesPaid), new Big(0));
+
+    const difference = totalFeesPaidToUkefForThePeriodInPaymentCurrency.minus(this.amountReceived).abs();
+    if (difference.gt(tolerance)) {
+      return false;
+    }
+    return true;
+  }
```

Comment:
How will this work and be used in the many to many case, e.g. payment 1 and payment 2 are both against fee record 1, fee record 2 and fee record 3. Then neither payment will match the total individually, they will need to be added to each other and then compared to the total of the reported fees. I'm tempted to say we should scrap this ticket and PR and include this within a PR of the wider context?

Reply:
I didn't think about that case, but you are right and that probably means we would need to have a separate function rather that a method on the entity. Unlike with the entity, we can't be sure that the `feeRecords` and `payments` that get passed to this hypothetical function are linked, so that would also require some additional checks.

Either way, I'm going to close this PR and just get this in with the `ADD_PAYMENT` event handler